src/merkle/simple_smt/mod.rs

```diff
@@ -0,0 +1,272 @@
+        let mut tree = Self { root, depth, store };
+        entries.try_for_each(|(key, leaf)| tree.insert_leaf(key, leaf))?;
+
+        Ok(tree)
+    }
+
+    /// Returns the root of this Merkle tree.
+    pub const fn root(&self) -> Word {
+        self.root
+    }
+
+    /// Returns the depth of this Merkle tree.
+    pub const fn depth(&self) -> u32 {
+        self.depth
+    }
+
+    /// Returns the set count of the keys of the leaves.
+    pub fn leaves_count(&self) -> usize {
+        self.store.leaves_count()
+    }
+
+    /// Returns a node at the specified key
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified depth is greater than the depth of the tree.
+    /// * The specified key does not exist
+    pub fn get_node(&self, depth: u32, key: u64) -> Result<Word, MerkleError> {
+        if depth == 0 {
+            Err(MerkleError::DepthTooSmall(depth))
+        } else if depth > self.depth() {
+            Err(MerkleError::DepthTooBig(depth))
+        } else if depth == self.depth() {
+            self.store.get_leaf_node(key)
+        } else {
+            let branch_node = self.store.get_branch_node(key, depth)?;
+            Ok(Rpo256::merge(&[branch_node.left, branch_node.right]).into())
+        }
+    }
+
+    /// Returns a Merkle path to the node at the specified key. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified key does not exist as a branch or leaf node
+    /// * The specified depth is greater than the depth of the tree.
+    pub fn get_path(&self, depth: u32, key: u64) -> Result<Vec<Word>, MerkleError> {
+        if depth == 0 {
+            return Err(MerkleError::DepthTooSmall(depth));
+        } else if depth > self.depth() {
+            return Err(MerkleError::DepthTooBig(depth));
+        } else if depth == self.depth() && !self.store.check_leaf_node_exists(key) {
+            return Err(MerkleError::InvalidIndex(self.depth, key));
+        }
+
+        let mut path = Vec::with_capacity(depth as usize);
+        let mut curr_key = key;
+        for n in (0..depth).rev() {
+            let parent_key = curr_key >> 1;
+            let parent_node = self.store.get_branch_node(parent_key, n)?;
+            let sibling_node = if curr_key & 1 == 1 {
+                parent_node.left
+            } else {
+                parent_node.right
+            };
+            path.push(sibling_node.into());
+            curr_key >>= 1;
+        }
+        Ok(path)
+    }
+
+    /// Returns a Merkle path to the node at the specified key. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified key does not exist as a leaf node
+    pub fn get_leaf_path(&self, key: u64) -> Result<Vec<Word>, MerkleError> {
+        if !self.store.check_leaf_node_exists(key) {
+            return Err(MerkleError::InvalidIndex(self.depth, key));
+        }
```

Comment:
Since we already do this check in `get_path()`, we can probably get rid of it here, no?

Reply:
True, should have been removed in the last commit. Fixed!